### PR TITLE
Fix #17 - Add destination reminders

### LINKED ISSUE #17: Destination reminders
Right now we only have arrival reminders for when a bus is coming to the
stop where you want to board.

Multiple users have requested the ability to have a reminder for when you
want to get off the bus.

This will be dependent on showing the trip schedule; we'll have to wait
until that API is stabilized and create a screen for it.

Labels: Milestone-1.5
Comment 3 by devin@twuni.org, Jun 13, 2011
A workaround for this issue could be to set a reminder for when the bus that you are on is scheduled to arrive at the stop where you want to get off.

### PR DESCRIPTION
**Pull request by barbeau**
~~*Please don't merge - work in progress*~~

This pull request implements destination reminders as proposed in #17.

You can activate a destination reminder using the below steps (also pictured in screenshots below):
1. Tapping on an arrival time for the bus you want to receive a destination alert for
1. Choose "Show trip status"
1. Long press on the stop you want to receive the destination alert for
1. Tap "Yes"

image

TODO:
- [x] Refactor from "*TAD*" to "*Nav*"
- [x] Fix unit test framework, based on master branch changes
- [x] Fix input for failing test `NavigationTest.testTrip2()`
- [x] Fix input for failing test `NavigationTest.testTrip1C()` - `java.lang.NumberFormatException: For input string: ""`
- [x] Fix input for failing test `NavigationTest.testTrip2C()` - java.lang.NumberFormatException: For input string: "1.468E+12"`
- [x] Fix failing test `NavigationTest.testTrip16()` - use correct input data
- [x] Improve algorithm to pass failing test `NavigationTest.testTrip3()`
- [x] Improve algorithm to pass failing test `NavigationTest.testTrip3C()`
- [x] Figure out what to do with test `NavigationTest.testTrip14()` - "This test is failing because the second to last stop in the map is not in the coordinate where the actual stop is located. Hence, it appears that the pull the cord is triggered to soon"
- [x] Currently "Pull the Cord now" is announced 10 times based on the code (and when running unit tests).  Check to see if that's cancelable by the user
- [x] Decrease navigation unit test execution time - For example, testTrip9 (1125 location fixes in `tad_trip9.csv`) currently takes 22 seconds, and testTrip13 (2766 location fixes in `tad_trip13.csv`) currently takes 58 seconds on my desktop machine.  It looks like `NavigationServiceProvider.updateInterface()` takes the most time to execute - with calls to that method commented out in checkProximityAll() testTrip9 takes 900ms to execute, and all tests execute in 22 seconds.  However, there is additional test overhead to set up tests that isn't reflected in this time - actual running time is more.  (I fixed this in https://github.com/OneBusAway/onebusaway-android/commit/a1657c443258ac49b1be83a75399cf2ced71080e)
- [x] Collect feedback data on algorithm (use Cloud Firestore?  No, use Firebase Storage instead) - https://firebase.google.com/docs/firestore/data-model
    * See also using LiveData/ViewModel and Firestore - https://firebase.googleblog.com/2017/12/using-android-architecture-components.html
    * https://proandroiddev.com/firebase-android-series-firestore-17e8951c574e
    * Documentation for only allowing users to access their own data - ~~https://youtu.be/rtoxRg-kbt0?t=5m29s~~ https://firebase.google.com/docs/firestore/security/get-started
    * Example storing location data - https://medium.com/google-cloud/gps-cellular-asset-tracking-using-google-cloud-iot-core-firestore-and-mongooseos-4dd74921f582
    * Security video - https://youtu.be/oFlHzF5U-HA
    * Secure the Firestore data - https://firebase.google.com/docs/firestore/quickstart#secure_your_data
- [x] Securing Firebase File storage - make sure files are writable but not readable by the app
- [x] Resolve the Firebase conflict with Embedded Social - https://github.com/Microsoft/EmbeddedSocial-Android-SDK/issues/95
- [x] Make sure temporary CSV files are stored in private, secure location - we should be writing only to internal storage - https://developer.android.com/training/data-storage/files
- [x] Remove default debug logging option from build config, replace it with user option (or something else?)
- [x] Settle on "alert" or "reminder" naming convention for customer-facing text.  Given that we already have arrival reminders, destination "reminders" probably make the most sense.
- [x] More testing with NavigationService - I'm not sure if a recent code change introduce some strange behavior in NavigationService.onLocationChanged(), but something definitely wasn't right.  I've changed it in 025ac8035b5184e4e083ba08c67bec4dddc4a261 but it needs more testing in the field to confirm that the LocationService is properly triggering the navProvider.
- [x] Improve "we found a better trip" notifications? (https://github.com/OneBusAway/onebusaway-android/issues/17#issuecomment-342275556) - **EDIT** New issue opened for this at https://github.com/OneBusAway/onebusaway-android/issues/973
- [x] Saved/scheduled trips? (https://github.com/OneBusAway/onebusaway-android/issues/17#issuecomment-343181149) - **EDIT** New issue opened for this at https://github.com/OneBusAway/onebusaway-android/issues/974
- [x] Support nav_trip*.csv test files that include extra newlines at the bottom (currently test crashes when trying to parse coordinate ID from the empty string) - **EDIT** New issue opened for this at https://github.com/OneBusAway/onebusaway-android/issues/975
- [x] Investigate duplicate records getting logged to CSV - nano-since-boot is being duplicated
- [x] Add ability to log stop closest to the user's real-time location when they start their trip (derived from stop they are looking at before setting the reminder, or from vehicle on the map compare user location to stops on map) - **EDIT** We're not going to do this at this stage due to privacy reasons and project timeline
- [x] Don't capture data until user is near "Get ready" radius
- [x] Update ES SDK version - https://github.com/Microsoft/EmbeddedSocial-Android-SDK/issues/113

UI feedback flow cases to handle:
- [x] User never responses to notification - Delete the file after a certain amount of time.
- [x] User responds "Yes" but doesn't want to include data - Log Yes via Firebase Analytics.  Delete file.
- [x] User responds "No" but doesn't want to include data - Log No via Firebase Analytics (including plain text feedback?). Delete file.
- [x] User responds "Yes" and they do want to include data - Figure out how to indicate file is yes or no (upload to separate server folder?). Upload CSV to [Firebase Cloud Storage](https://firebase.google.com/docs/storage/) and schedule via [WorkManager](https://developer.android.com/topic/libraries/architecture/workmanager/).  After upload delete file locally.
- [x] User responds "No" and they do want to include data - Figure out how to indicate file is yes or no (upload to separate server folder?). Upload CSV to [Firebase Cloud Storage](https://firebase.google.com/docs/storage/) and schedule via [WorkManager](https://developer.android.com/topic/libraries/architecture/workmanager/).  After upload delete file locally.

~~To handle above, likely need to add another database field that associates trip with CSV file so we know which one to upload after user gives feedback.~~

We will store plain text user feedback using the FirebaseAnalytics event parameters `CONTENT` - see https://firebase.google.com/docs/reference/android/com/google/firebase/analytics/FirebaseAnalytics.Param

### CODE REVIEW

**Comment by paulnabanita1** [comment]
@barbeau Thank you for the update. After fetching the latest commit, I am unable to run the test from the NavigationTest.java file. Can you please let me know if I have to make any other changes to run the tests?
Below is the error details :
org.gradle.api.tasks.TaskExecutionException: Execution failed for task ':onebusaway-android:compileObaGoogleDebugAndroidTestJavaWithJavac'
error: constructor PathLink in class PathLink cannot be applied to given types;
required: long,Location,Location,Location,String
found: <null>,Location,Location,String
reason: actual and formal argument lists differ in length****

**Comment by barbeau** [comment]
@paulnabanita1 Whoops!  Thanks for letting me know. I'll fix that shortly.

**Comment by barbeau** [comment]
@paulnabanita1 Ok, NavigationTest should be fixed in https://github.com/OneBusAway/onebusaway-android/pull/873/commits/5aef3aa95ac7086fba84798a54e1a6597069474b.  Could you please try it again?

**Comment by paulnabanita1** [comment]
@barbeau It is working now. Thank you.

**Comment by barbeau** [comment]
@paulnabanita1 I merged the master branch into the `nav` branch via https://github.com/OneBusAway/onebusaway-android/pull/873/commits/35598c9ac15fcbdaed327cfbe764e4e54855abc7, which includes the migration to Android X libraries instead of old support libraries.

Could you please do some testing to make sure that everything works as before, and that the merge didn't break anything?

Following this, you can start working on the new feedback notifications.  I noticed in NavigationServiceProvider line 664 that `NotificationCompat.Builder` is now deprecated, and we should be using notification channels as specified in https://developer.android.com/training/notify-user/channels.  Could you please find everywhere in the project where we are using notifications (i.e., search for the notification builder) , and update them to use channels?

Off the top of my head we use notifications for:
* Bus arrival reminders
* Trip plan changes based on real-time info
* Destination reminders (this PR)

You'll also need to check and see if the "Notification" settings in the OBA app settings still work (or there is at least some way to change them in system settings) after transitioning to using channels.

**Comment by paulnabanita1** [comment]
@barbeau Thank you for letting me know. I will test to check everything is working as before and let you know if I face any issues.

**Comment by barbeau** [comment]
@paulnabanita1 I've finished implementing Firebase analytics and merged that into the `nav` branch, so we can use that for submitting the user feedback "yes", "no", and plain text responses.

You can implemente this by adding a new method to the `ObaAnalytics` class that looks like the following:

~~~
    /**
     * Reports destination reminder feedback using Firebase
     * @param analytics Firebase singleton
     * @param wasGoodReminder true if the user responded that they got the reminder at the right time, or false if they responded that they did not get the reminder at the right time
     * @param feedbackText plain text feedback submitted by the user, or null if the user didn't enter any feedback text
     * @param fileName the name of the file that was uploaded that contains the data for this particular trip, or null if the user didn't upload data
     */
    public static void reportDestinationReminderFeedback(FirebaseAnalytics analytics, boolean wasGoodReminder, String feedbackText, String fileName) {
        if (!isAnalyticsActive()) {
            return;
        }
        Bundle bundle = new Bundle();
        bundle.putString(FirebaseAnalytics.Param.ITEM_ID, Application.get().getString(R.string.analytics_label_button_press_destination_reminder));
        if (wasGoodReminder) {
            bundle.putString(FirebaseAnalytics.Param.ITEM_VARIANT, Application.get().getString(R.string.analytics_label_destination_reminder_yes));
        } else {
           bundle.putString(FirebaseAnalytics.Param.ITEM_VARIANT, Application.get().getString(R.string.analytics_label_destination_reminder_no));
       }
        if (!isEmpty(feedbackText)) {
            bundle.putString(FirebaseAnalytics.Param.CONTENT, feedbackText);
        }
        if (!isEmpty(fileName)) {
            bundle.putString(FirebaseAnalytics.Param.ITEM_LOCATION_ID, fileName);
        }
        analytics.logEvent(FirebaseAnalytics.Event.SELECT_CONTENT, bundle);
    }
~~~

...and we'll need to add a new string to `donottranslate.xml` that looks like:

`<string name="analytics_label_button_press_destination_reminder_feedback">destination_reminder_feedback</string>`
`<string name="analytics_label_destination_reminder_yes">yes</string>`
`<string name="analytics_label_destination_reminder_no">no</string>`

Then, you can call this method accordingly when the user interacts with the feedback notification and activity.  Note that you'll need to pass in `null` for `feedbackText` if the user doesn't enter any free text response, and `null` for `fileName` if the user doesn't submit any data with the feedback.

If there is a way to get a complete URI to the uploaded file, instead of just using the local file name, that would be ideal, but I don't know if that's possible.

**Comment by paulnabanita1** [comment]
@barbeau  Thank you for the update. I will add the changes to the code and see if it works as expected. Also, I will have a look at whether it is possible to get a URI to the uploaded file and let you know.

File: onebusaway-android/src/main/java/org/onebusaway/android/nav/NavigationUploadWorker.java [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+                uploadTask.addOnFailureListener(new OnFailureListener() {
+                    @Override
+                    public void onFailure(@NonNull Exception exception) {
+                        Log.e(TAG, "Log upload failed");
+                    }
+                }).addOnSuccessListener(new OnSuccessListener<UploadTask.TaskSnapshot>() {
+                    @Override
+                    public void onSuccess(UploadTask.TaskSnapshot taskSnapshot) {
+                        Log.d(TAG, logFileName + " uploaded successful");
+                        boolean deleted = lFile.delete();
```
Comment (barbeau) [inline onebusaway-android/src/main/java/org/onebusaway/android/nav/NavigationUploadWorker.java]: @paulnabanita1 Actually, I think we're still missing the Firebase Analytics logging here (and when the user doesn't want to upload the file), right?

Comment (paulnabanita1) [reply onebusaway-android/src/main/java/org/onebusaway/android/nav/NavigationUploadWorker.java]: @barbeau We are yet to implement the Firebase analytics logging. I will open a new PR with the required changes.
